addon/internal-session.js

```diff
@@ -44,59 +45,87 @@ export default Ember.ObjectProxy.extend(Ember.Evented, {
+      this._callStoreAsync('restore').then((restoredContent) => {
+        let { authenticator } = (restoredContent.authenticated || {});
+        if (!!authenticator) {
+          delete restoredContent.authenticated.authenticator;
+          getOwner(this).lookup(authenticator).restore(restoredContent.authenticated).then((content) => {
+            this.set('content', restoredContent);
+            this._setup(authenticator, content).then(resolve, reject);
+          }, () => {
+            Ember.Logger.debug(`The authenticator "${authenticator}" rejected to restore the session - invalidating…`);
+            this.set('content', restoredContent);
+            this._clear().then(reject, reject);
+          });
+        } else {
+          delete (restoredContent || {}).authenticated;
           this.set('content', restoredContent);
-          this._setup(authenticator, content);
-          resolve();
-        }, () => {
-          Ember.Logger.debug(`The authenticator "${authenticator}" rejected to restore the session - invalidating…`);
-          this.set('content', restoredContent);
-          this._clear();
-          reject();
-        });
-      } else {
-        delete (restoredContent || {}).authenticated;
-        this.set('content', restoredContent);
-        this._clear();
-        reject();
-      }
+          this._clear().then(reject, reject);
+        }
+      }, () => {
+        this._clear().then(reject, reject);
+      });
     });
   },
 
+  _callStoreAsync(method, ...params) {
+    let result = this.store[method].apply(this.store, params);
+    if (typeof result === 'undefined' || typeof result.then === 'undefined') {
+      Ember.deprecate(`Ember Simple Auth: Synchronous stores have been deprecated. Make sure your custom store's ${method} method returns a promise.`, false, {
+        id: `ember-simple-auth.session-store.synchronous-${method}`,
+        until: '2.0.0'
+      });
+      return RSVP.Promise.resolve(result);
+    } else {
+      return result;
+    }
+  },
+
   _setup(authenticator, authenticatedContent, trigger) {
-    trigger = !!trigger && !this.get('isAuthenticated');
-    this.beginPropertyChanges();
-    this.setProperties({
-      isAuthenticated: true,
-      authenticator
+    return new RSVP.Promise((resolve, reject) => {
+      trigger = !!trigger && !this.get('isAuthenticated');
+      this.beginPropertyChanges();
+      this.setProperties({
+        isAuthenticated: true,
+        authenticator
+      });
+      Ember.set(this.content, 'authenticated', authenticatedContent);
+      this._bindToAuthenticatorEvents();
+      this._updateStore().then(() => {
+        this.endPropertyChanges();
+        if (trigger) {
+          this.trigger('authenticationSucceeded');
+        }
+        resolve();
+      }, () => {
+        this.setProperties({
+          isAuthenticated: false,
+          authenticator: null
+        });
+        Ember.set(this.content, 'authenticated', {});
```

Comment:
I think there must be a `this.endPropertyChanges();` after this line as well as otherwise it would remain in the state set by `beginPropertyChanges` above.